File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.go
type RandomizedSet struct {
    data map[int]struct{}
}


func Constructor() RandomizedSet {
    rand.Seed(time.Now().UnixNano())
    return RandomizedSet{
        data: make(map[int]struct{}),
    }
}


func (this *RandomizedSet) Insert(val int) bool {
    if _, ok := this.data[val]; !ok {
        this.data[val] = struct{}{}
        return true
    }
    return false
}


func (this *RandomizedSet) Remove(val int) bool {
    if _, ok := this.data[val]; ok {
        delete(this.data, val)
        return true
    }
    return false
}


func (this *RandomizedSet) GetRandom() int {
    idx := rand.Intn(len(this.data))
    now := 0
    for val := range this.data {
        if now == idx {
            return val
        }
        now++
    }
    return -1
}


/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */

File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.cpp
class RandomizedSet {
public:
    set <int> storage;
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if (storage.find(val) != storage.end()) {
            return false;
        }
        storage.insert(val);
        return true;
    }
    
    bool remove(int val) {
        if (storage.find(val) == storage.end()) {
            return false;
        }
        storage.erase(val);
        return true;
    }
    
    int getRandom() {
        int max = storage.size()-1;
        int min = 0;
        int range = max - min + 1;
        int num = rand() % range + min;
        return *next(storage.begin(), num);
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */